A `ring_exp` tactic for dealing with exponents in rings

Vierkantor opened the pull request:
The `ring_exp` tactic is an extension of `ring` that has support for exponents. In other words, it should normalize everything that `ring` can, but it can also recursively normalize any expression appearing as an exponent. (The representation used in `ring` is much more efficient when the exponents are large-ish numbers, so it still makes sense to keep `ring`.)

I have tested the effectivity of the tactic in two ways: writing a set of tests, and building mathlib with every call to `ring` replaced with `ring_exp`. The rebuild took approximately the same time as the original version and (of course) gave no errors.

I changed two proofs in `data.polynomial` to use `ring_exp` as a demonstration of its usefulness.

TO CONTRIBUTORS:

Make sure you have:

  * [x] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [x] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [x] for tactics:
     * [x] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [x] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)
  * [x] make sure definitions and lemmas are put in the right files
  * [x] make sure definitions and lemmas are not redundant

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

Vierkantor committed:
Test for ring_exp

Vierkantor committed:
Implement -a/b * -a/b = a/b * a/b

Vierkantor committed:
Hide extra information in the `ex` type in `ex_info`

Vierkantor committed:
Some attempts to make the proof returned by ring_exp shorter

Vierkantor committed:
Fix that ring_exp wouldn't handle pow that isn't monomial.has_pow

Vierkantor committed:
Some optimizations in ring_exp

Vierkantor committed:
Make all proofs explicit, halving execution time more or less

Vierkantor committed:
Cache `has_add` and `has_mul` instances for another 2x speedup

Vierkantor committed:
ring_exp can replace ring to compile mathlib

Vierkantor committed:
Revert `ring` to non-test version

Vierkantor committed:
Code cleanup and documentation

Vierkantor committed:
Revert the test changes to `linarith`

Vierkantor committed:
Undo the test changes to `ring2`

Vierkantor committed:
Whitespace cleanup

Vierkantor committed:
Fix overzealous error handling

Instead of catching any `fail` in eval, we just catch the operations that can
safely `fail` (i.e. `invert` and `negate`). This should make internal errors
visible again.

Vierkantor committed:
Fix the TODO's

Vierkantor committed:
Example use of ring_exp in data.polynomial

Vierkantor committed:
Check that `ring_exp` deals well with natural number subtraction

Vierkantor committed:
Fix incorrect docstring

Vierkantor committed:
Improve documentation

Vierkantor committed:
Small stylistic fixes

Vierkantor committed:
Fix slow behaviour on large exponents

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1419 @@
+/--
+  The `ring_exp_m` monad is used instead of `tactic` to store the context.
+-/
+meta def ring_exp_m (α : Type) : Type := reader_t context (state_t (list atom) tactic) α
```
```suggestion
@[derive [monad, alternative]]
meta def ring_exp_m (α : Type) : Type := reader_t context (state_t (list atom) tactic) α
```
This should replace the two instances below.

Vierkantor replied:
I get the following error if I try to apply the suggestion:
```
mathlib/src/tactic/ring_exp.lean:371:9: error: type mismatch at application
  monad (ring_exp_m a)
term
  ring_exp_m a
has type
  Type : Type 1
but is expected to have type
  Type ? → Type ? : Type (max (?+1) (?+1))
```
also if I remove the bound variable α:
```lean
@[derive [monad, alternative]]
meta def ring_exp_m : Type → Type := reader_t context (state_t (list atom) tactic)
```

robertylewis replied:
Ah, yeah. I see the delta derive handler isn't well suited to derive instances on pi types. There's not an easy solution there so we can just leave this.

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1419 @@
+
+meta def eval : expr → ring_exp_m (ex sum)
+| e@`(%%ps + %%qs) := do
+  ps' ← eval ps,
```
These lines can be written more concisely as `add <$> ps' <*> qs'` if monad glyphs are your thing. Not a necessary change though, it's a matter of taste. Same applies below.

Vierkantor replied:
Unfortunately, the type of `add` is `ex sum → ex sum → ring_exp_m (ex sum)`, so we really need a monadic bind here.

Vierkantor replied:
In commit f9b72160, I switched other terms to use applicative notation (where I could do so without introducing a λ term). I'm unsure which version I prefer if the other code is already in `do` notation, so I would be happy to revert that commit.

robertylewis replied:
Ah, I see. It's your call. There are no style guidelines about this and you'll see both in the library.

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1419 @@
+namespace tactic.interactive
+open interactive interactive.types lean.parser tactic tactic.ring_exp
+/-- Repeatedly apply `eval_simple` on (sub)expressions. -/
+meta def normalize (e : expr) : tactic (expr × expr) := do
```
This probably doesn't belong in `tactic.interactive`.

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1419 @@
+    tactic.interactive.exact ``(%%pf)
+  else fail "ring_exp failed to prove equality"
+
+/-- Tactic for solving equations of *commutative* (semi)rings,
```
```suggestion
/-- Tactic for normalizing expressions of *commutative* (semi)rings,
```
This version does more than solve equations, right?

robertylewis reviewed (commented):
Some very brief comments now. I'll write more tomorrow!

Vierkantor committed:
Add `ring_exp` to the default tactics

Vierkantor committed:
Use applicative notation where appropriate

Vierkantor committed:
The `ring_exp` tactic also does normalization

Co-Authored-By: Rob Lewis <Rob.y.lewis@gmail.com>

Vierkantor committed:
Move `normalize` from `tactic.interactive` to `ring_exp` namespace

Vierkantor committed:
Fix name collision between `equiv` in data.equiv.basic and `equiv` in `test/tactics.lean`

I just renamed the definition in `test/tactics.lean` to `my_equiv`
and the operator to `my≅`.

Vierkantor committed:
Fixes for the linter

Vierkantor committed:
Fix the usage of type classes for `sub_pf` and `div_pf`

Vierkantor committed:
Fix an additional linting error

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+  so division cancelling out is not true in general.
+
+  Multiplication of powers can be simplified a little bit further:
+  `2 ^ n * 2 ^ n = 4 ^ n := by ring_exp` could be implemented
```
As we discussed in person, this will come up soon enough. But It's reasonable to do this in a later PR.

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+  but `ring_exp` does if `ring_exp_eq` was not sufficient.
+
+  Atoms in fact represent equivalence classes of expressions,
+  modulo definitional equality.
```
`ring` and `linarith` have an option that controls how much unfolding they do to test definitional equality. e.g.
`example (x y : ℕ) : x + id y = y + id x := by ring!` works but not without the `!`. It's usually better to default to the weaker test. I'm guessing I can find where this is done later in this file, so I'll comment again when I get there.

Vierkantor replied:
Done!

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+
+section expression
+/-!
+  ## `expression` section
```
```suggestion
  ### `expression` section
```
Changing my mind from before. Since we use h2 headers in the header docs, using h3 internally seems better.

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+  or using a different meta representation of numerals.
+-/
+@[derive decidable_eq]
+structure coeff : Type := (value {} : ℚ)
```
```suggestion
structure coeff : Type := (value : ℚ)
```
There are no variables, right? I don't think this does anything.

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+/--
+  The `ex` type is structured according to a parameter `et : ex_type`.
+
+  Instead of allowing each operand to have arbitrary subexpressions as argument,
```
I'm a little confused by this comment. It seems to be describing `ex`, not `ex_type`, no?

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+  (For example, the correct equality check of `expr` is `is_def_eq : expr → expr → tactic unit`.)
+-/
+meta inductive ex : ex_type → Type
+| zero  {} (info : ex_info) : ex sum
```
```suggestion
| zero (info : ex_info) : ex sum
```
I don't think any of these should be necessary either. Or am I misunderstanding?

Vierkantor replied:
I forgot to remove them after removing parameters to `ex`, so they are indeed unnecessary!

Vierkantor committed:
Optimization: we don't need norm_num to determine `x * 1 = x`

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+meta def ex.pretty {et : ex_type} (ps : ex et) : expr := ps.info.pretty
+/--
+  Return the normalisation proof of the given expression.
+  If the proof is `refl`, we give `none` instead,
```
This is only done when the expression is syntactically equal to its normal form, right? You're not using `refl` to do any computation?

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+
+section operations
+/-!
+  ## `operations` section
```
```suggestion
  ### `operations` section
```

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+
+  This type is used in the function `add_overlap`.
+  In order to deal with equations of the form `a * 2 + a = 3 * a`,
+  the `add` function will add up overlapping producs,
```
```suggestion
  the `add` function will add up overlapping products,
```

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+  the `add` function will add up overlapping producs,
+  turning `a * 2 + a` into `a * 3`.
+  We need to distinguish `a * 2 + a` from `a * 2 + b` in order to do this,
+  and the `overlap` type carrues the information on how it overlaps.
```
```suggestion
  and the `overlap` type carries the information on how it overlaps.
```

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+meta def resolve_atoms (a : expr) : list atom → ℕ → ring_exp_m (atom × list atom)
+| [] n := let atm : atom := ⟨a, n⟩ in pure (atm, [atm])
+| bas@(b :: as) n := do
+  (lift $ is_def_eq a b.value >> pure (b , bas)) <|> do
```
Back to my comment toward the top. `is_def_eq` takes an optional reducibility parameter. The default is `semireducible`. You should take an optional flag in the `ring_exp` front end (`ring_exp!`) and thread the flag to here. If the flag is present, use `semireducible`. Otherwise use `reducible`.

Vierkantor replied:
Done!

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1430 @@
+
+section wiring
+/-!
+  ## `wiring` section
```
```suggestion
  ### `wiring` section
```

Vierkantor committed:
Improve documentation of `test/ring_exp.lean`

Vierkantor committed:
Rename `resolve_atoms` to `resolve_atom_aux` for clarity

robertylewis reviewed (commented):
This looks really good. As I type this, you're fixing (or have fixed) the performance issue we discussed. A lot of these are very minor; most important is the reducibility setting for matching atoms.

I think we should replace `ring` with this in a follow up PR. I'd like to test how this behaves on the problems from `linarith`, which don't have any exponents. (For the public record, Tim just showed me changes that make `ring_exp` significantly faster than `ring` on some examples.)

Vierkantor committed:
Small stylistic fixes

Vierkantor committed:
Remove unneccessary hidden fields to `ex`

Vierkantor committed:
Control how much unfolding `ring_exp` does by putting a `!` after it

robertylewis commented on src/tactic/ring_exp.lean:
```diff
@@ -0,0 +1,1452 @@
+  else fail "ring_exp failed to prove equality"
+
+/-- Tactic for normalizing expressions of *commutative* (semi)rings,
+    allowing variables in the exponent.
```
Could you document the `!` flag here, in `ring_exp_eq`, and in `tactics.md`?

Vierkantor committed:
Reword comment for `ex_type`

Vierkantor committed:
Use `ring_exp!` to deal with `(n : ℕ) + 1 - 1 = n`

Vierkantor committed:
Document the `!` flag for `ring`, `ring_exp` and `ring_exp_eq`

Vierkantor committed:
Get rid of searching for another cached instance

Vierkantor committed:
Fix `ring_exp` failing on terms on the form `0^succ (succ ...)`

digama0 reviewed (approved):
TODO(me): profiling and optimization
